python wrappers

Thomas-Ulrich opened the pull request:
add python wrapper for easi.
Example of use:

```python
import numpy as np
import easi

coords = np.random.rand(10,3)
region = np.random.randint(1,3, 10)
print(easi.evaluate_model(coords, region, ["rho", "lambda"], "test.yaml"))
```
output with (https://github.com/SeisSol/easi/blob/master/examples/1_groups.yaml) and regions `[2 1 1 2 1 1 1 2 1 1]`:
```
{'rho': array([2700., 2600., 2600., 2700., 2600., 2600., 2600., 2700., 2600.,
       2600.]), 'lambda': array([3.24038016e+10, 2.08000000e+10, 2.08000000e+10, 3.24038016e+10,
       2.08000000e+10, 2.08000000e+10, 2.08000000e+10, 3.24038016e+10,
       2.08000000e+10, 2.08000000e+10])}
```

Will then be integrated to spack as:

https://github.com/Thomas-Ulrich/spack/tree/thomas/easi_pybind

sebwolf-de commented on CMakeLists.txt:
```diff
@@ -57,6 +62,23 @@ if(ASAGI)
 endif()
 
+if (PYTHON_BINDINGS)
+    set(PYBIND11_VER 2.11.1)
```
Please add `set(PYBIND11_FINDPYTHON ON)` for compability with new cmake versions, c.f. https://github.com/pybind/pybind11/blob/master/tools/FindPythonLibsNew.cmake#L102-L108

sebwolf-de commented on python_bindings/README.md:
```diff
@@ -0,0 +1,33 @@
+# A python wrapper for easi
+
+Install with `-DPYTHON_BINDINGS=ON` cmake option.
```
Note that ASAGI and ImpalaJIT have to be compiled with `cmake .. -DCMAKE_CXX_FLAGS="-fPIC"`.

sebwolf-de commented on python_bindings/README.md:
````diff
@@ -0,0 +1,33 @@
+The module can then be used as follows:
+
+```python
+import numpy as np
+import easi
+
+coords = np.random.rand(10,3)
+region = np.random.randint(1,3, 10)
+print(easi.evaluate_model(coords, region, ["rho", "lambda"], "test.yaml"))
````
When I try this, I get this error:
```
Traceback (most recent call last):
  File "/opt/easi/pytest.py", line 2, in <module>
    import easi
ModuleNotFoundError: No module named 'easi'
```

Thomas-Ulrich replied:
how did you install it? with spack or manually?
If you install it manually, you need to do:
```export PYTHONPATH=~/SeisSol/easi/build/python_bindings:$PYTHONPATH``` or similar.
Note that the version of python (e.g. 3.10 `easi.cpython-310-x86_64-linux-gnu.so`) has to be the same as the python version that you use for loading the module.
(this may happen if the python in spack ~/.spack/packages.yaml is the system python, and your python is a conda installed python).

sebwolf-de commented on python_bindings/README.md:
```diff
@@ -0,0 +1,33 @@
+# A python wrapper for easi
```
Could you add this README to the RTD documentation? https://github.com/SeisSol/easi/tree/master/doc

sebwolf-de reviewed (commented):
Very nice feature! Unfortunately I could not run it on my machine :(

sebwolf-de commented on python_bindings/README.md:
````diff
@@ -0,0 +1,33 @@
+Add the library to your path:
+
+```bash
+export PYTHONPATH=your_build_folder>python_bindings:$PYTHONPATH
````
Replace `>` with `/`

Thomas-Ulrich commented:
I addressed the comments. I also fixed this error in compiling the unit tests with gcc12.2:
```
In file included from /opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/bits/stl_algobase.h:64,
                 from /opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/memory:63,
                 from /home/ulrich/SeisSol/easi/include/easi/util/Matrix.h:7,
                 from /home/ulrich/SeisSol/easi/include/easi/Query.h:4,
                 from /home/ulrich/SeisSol/easi/tests/easitest.h:5,
                 from /home/ulrich/SeisSol/easi/tests/0_constant.cpp:1:
/opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/bits/stl_pair.h: In instantiation of ‘struct std::pair<int, std::array<double, 3> >’:
/home/ulrich/SeisSol/easi/tests/0_constant.cpp:6:39:   required from here
/opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/bits/stl_pair.h:192:11: error: ‘std::pair<_T1, _T2>::second’ has incomplete type
  192 |       _T2 second;                ///< The second member
      |           ^~~~~~
In file included from /opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/bits/unique_ptr.h:36,
                 from /opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/memory:76:
/opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/tuple:1595:45: note: declaration of ‘struct std::array<double, 3>’
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
In file included from /opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/bits/stl_pair.h:60:
/opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/type_traits: In instantiation of ‘struct std::is_copy_assignable<std::array<double, 3> >’:
/opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/type_traits:167:12:   required from ‘struct std::__and_<std::is_copy_assignable<int>, std::is_copy_assignable<std::array<double, 3> > >’
/opt/software/gcc/12.2.0/lib/gcc/x86_64-pc-linux-gnu/12.2.0/include/c++/bits/stl_pair.h:568:7:   required from ‘struct std::pair<int, std::array<double, 3> >’
/home/ulrich/SeisSol/easi/tests/0_constant.cpp:6:39:   required from here
```

Thomas-Ulrich commented:
Should I change the version to 1.3.0?

sebwolf-de commented:
> Should I change the version to 1.3.0?

Yes, that's a good idea.
